archive/zip_archiver.go

```diff
@@ -84,7 +107,38 @@ func checkMatch(fileName string, excludes []string) (value bool) {
 	return false
 }
 
-func (a *ZipArchiver) ArchiveDir(indirname string, excludes []string) error {
+// The basic file header is very simple. The UncompressedSize logic is not a real-world use case
+// in this context, but "640K ought to be enough for anybody".
+//
+// For reference, see golang/src/archive/zip/struct.go.
+func prepareEmptyHeader(content []byte, infilename string) *zip.FileHeader {
+	fh := &zip.FileHeader{
+		Name:               filepath.ToSlash(infilename),
+		UncompressedSize64: uint64(len(content)),
+	}
+
+	if fh.UncompressedSize64 > uint32max {
+		fh.UncompressedSize = uint32max
+	} else {
+		fh.UncompressedSize = uint32(fh.UncompressedSize64)
+	}
+
+	return fh
+}
+
+// Normalize the fields:
+//
+// - no compression, so the compressed stream is essentially a copy;
+// - fixed date;
+// - fixed file permissions.
+//
+func normalizeCompressingFile(fh *zip.FileHeader) {
+	fh.Method = zip.Store
+	fh.SetModTime(time.Date(1981, 4, 10, 0, 0, 0, 0, time.UTC))
+	fh.SetMode(0644)
```

Comment:
What is the file is executable?

See here for an example: https://github.com/terraform-providers/terraform-provider-archive/pull/41/files#diff-572809c26c1b9c23a665ebaea5dd7ea8R70

Reply:
I'm not familiar with executing binaries inside a Lambda functions; based on a brief research, Lambda resets the permissions of the archived files on mount - see, for example:

- https://www.reddit.com/r/aws/comments/7k0txe/how_can_i_give_executables_inside_a_lambda
- https://stackoverflow.com/questions/48194886/update-permissions-on-aws-lambda-file-to-make-it-executable

It's not 100% clear from those resources where their problem is - it could be the zip library, or Lambda resetting the permissions when unarchiving; if it's the latter case, managing executable bit would not yield any effect.

Do you have direct experience on this use case, so that I know with certainty what's the exact behavior?